### PR TITLE
Fix double grad hang bug

### LINKED ISSUE #33989: paddle.grad for large models is very slow
- 版本、环境信息：
    1）PaddlePaddle版本：2.0.1
    2）CPU：请提供CPU型号，MKL/OpenBlas/MKLDNN/等数学库的使用情况
    3）GPU：CUDA 10.2.89, CUDNN7.6.5
    4）系统环境：Ubuntu18.04，Python 3.7.0
 
--- 

简单来说，用Vision Transformer的模型时，我想用`paddle.grad`计算输出对于输入图片的gradient，耗时25s左右；即使是计算 输出 对于 下述代码中`out`的 gradient，耗时依然在23-25s左右。
同样的代码，模型换成resnet50，耗时0.1s。

麻烦看一下是什么原因~

```python
import paddle
from interpretdl.data_processor.readers import preprocess_inputs
from datetime import datetime

img_path = 'assets/catdog.png'
# read the image and scale to [0, 1] with substract mean and divided by std.
imgs, data = preprocess_inputs(img_path, [224, 224, 3])  
print(imgs.shape, data.shape)  # (1, 224, 224, 3) (1, 3, 224, 224)

# from paddle.vision.models import resnet50
# paddle_model = resnet50(pretrained=True)

from assets.vision_transformer import ViT_small_patch16_224
# https://github.com/PaddlePaddle/PaddleClas/blob/release/2.2/ppcls/arch/backbone/model_zoo/vision_transformer.py
paddle_model = ViT_small_patch16_224()
MODEL_PATH = 'assets/ViT_small_patch16_224_pretrained.pdparams'
paddle_model.set_dict(paddle.load(MODEL_PATH))

data = paddle.to_tensor(data)
data.stop_gradient = False
out = paddle_model(data)
preds = paddle.argmax(out, axis=1)
label_onehot = paddle.nn.functional.one_hot(
                    paddle.to_tensor(preds), num_classes=out.shape[1])
target = paddle.sum(out * label_onehot, axis=1)

now = datetime.now()
g = paddle.grad(outputs=target, inputs=data)[0]
print(datetime.now() - now)  # 0:00:23.528224
```

### PR DESCRIPTION
**Pull request by sneaxiy**
### PR types
Bug fixes

### PR changes
Others

### Describe
Avoid to transverse duplicate `GradOpNode`s when pruning the backward graph.

Fix https://github.com/PaddlePaddle/Paddle/issues/33989 .

### CODE REVIEW

**Review by wzzju — APPROVED** [review]
LGTM
